File: python/idmix/alphabet.py
"""基于自定义字符表的 RadixCodec（默认 idmix 文本层）。"""

from __future__ import annotations

import struct


class RadixCodec:
    """使用自定义字符表（Base-N）的二进制↔文本编解码器。"""

    def __init__(self, alphabet: str) -> None:
        chars = list(alphabet)
        if len(chars) < 2:
            raise ValueError("alphabet must have at least 2 unique characters")
        self.base = len(chars)
        self.chars = chars
        self._from_custom: dict[str, int] = {}
        for i, ch in enumerate(chars):
            if ch in self._from_custom:
                raise ValueError(f"alphabet contains duplicate character {ch!r}")
            self._from_custom[ch] = i

    @classmethod
    def new(cls, alphabet: str) -> RadixCodec:
        return cls(alphabet)

    def alphabet(self) -> str:
        return "".join(self.chars)

    def encode(self, data: bytes) -> str:
        if len(data) == 0:
            return self.chars[0]
        wrapped = struct.pack(">H", len(data)) + data
        n = int.from_bytes(wrapped, "big")
        return self._int_to_string(n)

    def decode(self, s: str) -> bytes:
        if not s:
            raise ValueError("empty string")
        n = self._string_to_int(s)
        raw = n.to_bytes((n.bit_length() + 7) // 8 or 1, "big")
        if raw == b"\x00" and n == 0:
            raw = b""
        for pad in (0, 1):
            buf = b"\x00" * pad + raw
            if len(buf) < 2:
                continue
            data_len = struct.unpack(">H", buf[:2])[0]
            if len(buf) != 2 + data_len:
                continue
            return buf[2:]
        raise ValueError("invalid encoded data length")

    def _int_to_string(self, n: int) -> str:
        if n == 0:
            return self.chars[0]
        chars: list[str] = []
        base = self.base
        while n > 0:
            n, rem = divmod(n, base)
            chars.append(self.chars[rem])
        return "".join(reversed(chars))

    def _string_to_int(self, s: str) -> int:
        n = 0
        for ch in s:
            if ch not in self._from_custom:
                raise ValueError(f"invalid character {ch!r}")
            n = n * self.base + self._from_custom[ch]
        return n

File: golang/alphabet.go
// alphabet.go 实现基于自定义字符表的 RadixCodec（默认 idmix 文本层）。
//
// 编码策略：在原始数据前附加 2 字节大端长度前缀，整体视为一个大整数，
// 再按自定义字符表做进制转换（类似无填充的 Base-N）。
package idmix

import (
	"encoding/binary"
	"errors"
	"fmt"
	"math/big"
)

var errNilCodecFunc = errors.New("codec function is nil")

// RadixCodec 使用自定义字符表（Base-N）的二进制↔文本编解码器。
type RadixCodec struct {
	base       int
	chars      []rune
	fromCustom map[rune]int
}

// NewRadixCodec 根据字符表创建 RadixCodec。
func NewRadixCodec(alphabet string) (*RadixCodec, error) {
	runes := []rune(alphabet)
	if len(runes) < 2 {
		return nil, errors.New("alphabet must have at least 2 unique characters")
	}
	rc := &RadixCodec{
		base:       len(runes),
		chars:      runes,
		fromCustom: make(map[rune]int, len(runes)),
	}
	for i, r := range runes {
		if _, ok := rc.fromCustom[r]; ok {
			return nil, fmt.Errorf("alphabet contains duplicate character %q", r)
		}
		rc.fromCustom[r] = i
	}
	return rc, nil
}

// Alphabet 返回字符表字符串。
func (rc *RadixCodec) Alphabet() string {
	return string(rc.chars)
}

// Base 返回进制基数。
func (rc *RadixCodec) Base() int {
	return rc.base
}

func (rc *RadixCodec) Encode(data []byte) (string, error) {
	if len(data) == 0 {
		return string(rc.chars[0]), nil
	}
	wrapped := make([]byte, 2+len(data))
	binary.BigEndian.PutUint16(wrapped, uint16(len(data)))
	copy(wrapped[2:], data)
	n := new(big.Int).SetBytes(wrapped)
	return rc.intToString(n), nil
}

func (rc *RadixCodec) Decode(s string) ([]byte, error) {
	if s == "" {
		return nil, errors.New("empty string")
	}
	n, err := rc.stringToInt(s)
	if err != nil {
		return nil, err
	}
	raw := n.Bytes()
	for pad := 0; pad <= 1; pad++ {
		buf := make([]byte, pad+len(raw))
		copy(buf[pad:], raw)
		if len(buf) < 2 {
			continue
		}
		dataLen := int(binary.BigEndian.Uint16(buf[:2]))
		if len(buf) != 2+dataLen {
			continue
		}
		return buf[2:], nil
	}
	return nil, errors.New("invalid encoded data length")
}

func (rc *RadixCodec) intToString(n *big.Int) string {
	if n.Sign() == 0 {
		return string(rc.chars[0])
	}
	base := big.NewInt(int64(rc.base))
	zero := big.NewInt(0)
	rem := new(big.Int)
	chars := make([]rune, 0, 32)
	for n.Cmp(zero) > 0 {
		n.DivMod(n, base, rem)
		chars = append(chars, rc.chars[rem.Int64()])
	}
	for i, j := 0, len(chars)-1; i < j; i, j = i+1, j-1 {
		chars[i], chars[j] = chars[j], chars[i]
	}
	return string(chars)
}

func (rc *RadixCodec) stringToInt(s string) (*big.Int, error) {
	n := big.NewInt(0)
	base := big.NewInt(int64(rc.base))
	for _, r := range s {
		idx, ok := rc.fromCustom[r]
		if !ok {
			return nil, fmt.Errorf("invalid character %q", r)
		}
		n.Mul(n, base)
		n.Add(n, big.NewInt(int64(idx)))
	}
	return n, nil
}
